File: 31-next-permutation/next-permutation.ts
/**
 Do not return anything, modify nums in-place instead.
 */
function nextPermutation(nums: number[]): void {
    const length: number = nums.length

    let pivotIndex: number = length - 2
    while (pivotIndex >= 0 && nums[pivotIndex] >= nums[pivotIndex + 1]) {
        pivotIndex--
    }

    if (pivotIndex >= 0) {
        let swapIndex: number = length - 1
        while (nums[swapIndex] <= nums[pivotIndex]) {
            swapIndex--
        }

        [nums[pivotIndex], nums[swapIndex]] = [nums[swapIndex], nums[pivotIndex]]
    }

    let left: number = pivotIndex + 1
    let right: number = length - 1

    while (left < right) {
        [nums[left], nums[right]] = [nums[right], nums[left]]
        left++
        right--
    }
};

File: 31-next-permutation/next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        pivot = None
        for i in range(len(nums) - 1, 0, -1):
            if nums[i] > nums[i-1]:
                pivot = i - 1
                break
        if pivot == None:
            nums.reverse()
            return
        
        swap = len(nums) - 1
        while nums[swap] <= nums[pivot]:
            swap -= 1
        
        nums[swap], nums[pivot] = nums[pivot], nums[swap]

        nums[pivot+1:] = reversed(nums[pivot+1:])
        return nums